File: src/algorithms/searching.js
export function* linearSearch({ arr, target }) {
  if (!arr || arr.length === 0) {
    yield {
      data: [],
      comparing: [],
      active: [],
      sorted: [],
      message: 'Error: Array is empty',
      pseudocodeLine: 0,
      phase: 'error',
      pointers: {},
    };
    return;
  }

  yield {
    data: [...arr],
    comparing: [],
    active: [],
    sorted: [],
    message: `Starting linear search for target ${target} in array of ${arr.length} elements`,
    pseudocodeLine: 0,
    phase: 'start',
    pointers: {},
  };

  for (let i = 0; i < arr.length; i++) {
    yield {
      data: [...arr],
      comparing: [i],
      active: [],
      sorted: [],
      message: `Comparing arr[${i}] = ${arr[i]?.val ?? arr[i]} with target ${target}`,
      pseudocodeLine: 2,
      phase: 'comparing',
      pointers: {},
    };

    if ((arr[i]?.val ?? arr[i]) === target) {
      yield {
        data: [...arr],
        comparing: [],
        active: [i],
        sorted: [],
        message: `Found target ${target} at index ${i}!`,
        pseudocodeLine: 3,
        phase: 'found',
        pointers: {},
      };
      return;
    }
  }

  yield {
    data: [...arr],
    comparing: [],
    active: [],
    sorted: [],
    message: `Target ${target} not found — exhausted all ${arr.length} elements`,
    pseudocodeLine: 4,
    phase: 'not-found',
    pointers: {},
  };
}

export function* binarySearch({ arr, target }) {
  if (!arr || arr.length === 0) {
    yield {
      data: [],
      comparing: [],
      active: [],
      sorted: [],
      message: 'Error: Array is empty',
      pseudocodeLine: 0,
      phase: 'error',
      pointers: {},
    };
    return;
  }

  let low = 0;
  let high = arr.length - 1;
  const eliminated = new Set();

  yield {
    data: [...arr],
    comparing: [],
    active: [],
    sorted: [],
    message: `Starting binary search for target ${target}. Setting LOW=0, HIGH=${high}`,
    pseudocodeLine: 1,
    phase: 'init',
    pointers: { LOW: low, HIGH: high },
    eliminated: [],
  };

  while (low <= high) {
    const mid = Math.floor((low + high) / 2);

    yield {
      data: [...arr],
      comparing: [],
      active: [mid],
      sorted: [],
      message: `Calculate MID = floor((${low} + ${high}) / 2) = ${mid}`,
      pseudocodeLine: 3,
      phase: 'calculate-mid',
      pointers: { LOW: low, MID: mid, HIGH: high },
      eliminated: [...eliminated],
    };

    yield {
      data: [...arr],
      comparing: [mid],
      active: [],
      sorted: [],
      message: `Comparing arr[${mid}] = ${arr[mid]?.val ?? arr[mid]} with target ${target}`,
      pseudocodeLine: 4,
      phase: 'comparing',
      pointers: { LOW: low, MID: mid, HIGH: high },
      eliminated: [...eliminated],
    };

    if ((arr[mid]?.val ?? arr[mid]) === target) {
      yield {
        data: [...arr],
        comparing: [],
        active: [mid],
        sorted: [],
        message: `Found target ${target} at index ${mid}!`,
        pseudocodeLine: 5,
        phase: 'found',
        pointers: { LOW: low, MID: mid, HIGH: high },
        eliminated: [...eliminated],
      };
      return;
    } else if ((arr[mid]?.val ?? arr[mid]) < target) {
      for (let e = low; e <= mid; e++) eliminated.add(e);
      low = mid + 1;

      yield {
        data: [...arr],
        comparing: [],
        active: [],
        sorted: [],
        message: `arr[${mid}] = ${arr[mid]?.val ?? arr[mid]} < ${target} — eliminate left half, set LOW = ${low}`,
        pseudocodeLine: 6,
        phase: 'eliminate-left',
        pointers: { LOW: low, HIGH: high },
        eliminated: [...eliminated],
      };
    } else {
      for (let e = mid; e <= high; e++) eliminated.add(e);
      high = mid - 1;

      yield {
        data: [...arr],
        comparing: [],
        active: [],
        sorted: [],
        message: `arr[${mid}] = ${arr[mid]?.val ?? arr[mid]} > ${target} — eliminate right half, set HIGH = ${high}`,
        pseudocodeLine: 8,
        phase: 'eliminate-right',
        pointers: { LOW: low, HIGH: high },
        eliminated: [...eliminated],
      };
    }
  }

  yield {
    data: [...arr],
    comparing: [],
    active: [],
    sorted: [],
    message: `Target ${target} not found — LOW (${low}) > HIGH (${high}), search space exhausted`,
    pseudocodeLine: 9,
    phase: 'not-found',
    pointers: { LOW: low, HIGH: high },
    eliminated: [...eliminated],
  };
}

File: backend/routes/searching.py
from flask import Blueprint, request, jsonify

searching_bp = Blueprint('searching', __name__)

def create_frame(data, comparing=None, found=None, active=None, eliminated=None, pointers=None, message='', pseudocodeLine=0, phase=''):
    return {
        "data": [d for d in data],
        "comparing": comparing or [],
        "found": found or [],
        "active": active or [],
        "eliminated": eliminated or [],
        "pointers": pointers or {},
        "message": message,
        "pseudocodeLine": pseudocodeLine,
        "phase": phase
    }

def extract_val(item):
    if isinstance(item, dict) and 'val' in item:
        return item['val']
    return item

@searching_bp.route('/linear', methods=['POST'])
def linear_search():
    req = request.get_json()
    arr = req.get('array', [])
    target = req.get('target')
    frames = []

    if target is None:
        frames.append(create_frame(arr, message='No target specified', phase='done'))
        return jsonify({"frames": frames})

    if not arr:
        frames.append(create_frame(arr, message='Array is empty', phase='done'))
        return jsonify({"frames": frames})

    a = arr.copy()
    n = len(a)
    
    frames.append(create_frame(a, message=f'Starting linear search for {target}', pseudocodeLine=1, phase='start'))

    for i in range(n):
        val = extract_val(a[i])
        frames.append(create_frame(a, comparing=[i], message=f'Checking index {i}: is {val} == {target}?', pseudocodeLine=2, phase='compare'))
        
        if val == target:
            frames.append(create_frame(a, found=[i], message=f'Target {target} found at index {i}!', pseudocodeLine=3, phase='found'))
            return jsonify({"frames": frames})

    frames.append(create_frame(a, message=f'Target {target} not found in the array.', pseudocodeLine=4, phase='not-found'))
    return jsonify({"frames": frames})


@searching_bp.route('/binary', methods=['POST'])
def binary_search():
    req = request.get_json()
    arr = req.get('array', [])
    target = req.get('target')
    frames = []

    if target is None:
        frames.append(create_frame(arr, message='No target specified', phase='done'))
        return jsonify({"frames": frames})

    if not arr:
        frames.append(create_frame(arr, message='Array is empty', phase='done'))
        return jsonify({"frames": frames})

    a = arr.copy()
    
    left = 0
    right = len(a) - 1
    eliminated = []

    def get_pointers(l, r, m=None):
        p = {"LOW": l, "HIGH": r}
        if m is not None:
            p["MID"] = m
        return p

    frames.append(create_frame(a, active=list(range(left, right + 1)), pointers=get_pointers(left, right), message=f'Starting binary search for {target}. Array must be sorted.', pseudocodeLine=1, phase='start'))

    while left <= right:
        mid = (left + right) // 2
        val = extract_val(a[mid])
        
        frames.append(create_frame(a, comparing=[mid], active=list(range(left, right + 1)), eliminated=eliminated.copy(), pointers=get_pointers(left, right, mid), message=f'Left={left}, Right={right}. Checking mid={mid} (value: {val})', pseudocodeLine=3, phase='compare'))

        if val == target:
            frames.append(create_frame(a, found=[mid], active=list(range(left, right + 1)), eliminated=eliminated.copy(), pointers=get_pointers(left, right, mid), message=f'Target {target} found at index {mid}!', pseudocodeLine=4, phase='found'))
            return jsonify({"frames": frames})
        
        if val < target:
            eliminated.extend(range(left, mid + 1))
            frames.append(create_frame(a, active=list(range(mid + 1, right + 1)), eliminated=eliminated.copy(), pointers=get_pointers(left, right, mid), message=f'{val} < {target}. Discarding left half (indices {left} to {mid}).', pseudocodeLine=6, phase='discard-left'))
            left = mid + 1
        else:
            eliminated.extend(range(mid, right + 1))
            frames.append(create_frame(a, active=list(range(left, mid)), eliminated=eliminated.copy(), pointers=get_pointers(left, right, mid), message=f'{val} > {target}. Discarding right half (indices {mid} to {right}).', pseudocodeLine=8, phase='discard-right'))

    frames.append(create_frame(a, active=[], eliminated=eliminated.copy(), message=f'Target {target} not found in the array.', pseudocodeLine=9, phase='not-found'))
    return jsonify({"frames": frames})
